### PR TITLE
SCANGRADLE-148 Gradle scanner scans some files in hidden folders

### PR DESCRIPTION
**Pull request by ADarko22**
This changes are on top of https://github.com/SonarSource/sonar-scanner-gradle/pull/226 that should be merged first

### CODE REVIEW

File: src/main/java/org/sonarqube/gradle/SourceCollector.java [hunk 1]
```diff
@@ -101,19 +112,27 @@ public SourceCollector(Set<Path> existingSources, Set<Path> directoriesToIgnore,
...
+      if (relativePath.toFile().isHidden()) {
+        return true;
+      }
+      relativePath = relativePath.getParent();
+    }
+    return false;
+  }
+
+  private boolean isChildOrGrandChildOfRoot(Path path) {
+    return path.getParent() != null && (path.getParent().equals(root) || path.getParent().getParent().equals(root));
```
Comment (alban-auzeill) [inline src/main/java/org/sonarqube/gradle/SourceCollector.java:135]: `getParent().getParent()` may be null
```java
    Path root = Path.of("a");
    Path path = Path.of("b/c");
    path.getParent().getParent().equals(root); // NullPointerException
 ```
You can flip equals `root.equals(path.getParent().getParent())`

File: src/main/java/org/sonarqube/gradle/SourceCollector.java [hunk 2]
```diff
@@ -92,7 +98,12 @@ public Set<Path> getCollectedSources() {
 
   private final Set<Path> collectedSources = new HashSet<>();
 
-  public SourceCollector(Set<Path> existingSources, Set<Path> directoriesToIgnore, Set<Path> excludedFiles, boolean shouldCollectJavaAndKotlinSources) {
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public SourceCollector(Path root, Set<Path> existingSources, Set<Path> directoriesToIgnore, Set<Path> excludedFiles, boolean shouldCollectJavaAndKotlinSources) {
```
Comment (dorian-burihabwa-sonarsource) [inline src/main/java/org/sonarqube/gradle/SourceCollector.java:105]: If we create an object using the builder pattern, is there a reason to keep the constructor public?

File: integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java [hunk 3]
```diff
@@ -354,7 +354,8 @@ public void testScanAllOnMultiModuleWithSubModulesProjectCollectsTheExpectedSour
         baseDir.resolve("module/submodule/submoduleScript.sh").toString(),
         baseDir.resolve("gradlew").toString(),
         baseDir.resolve("gradlew.bat").toString(),
-        baseDir.resolve("gradle/wrapper/gradle-wrapper.properties").toString()
+        baseDir.resolve("gradle/wrapper/gradle-wrapper.properties").toString(),
+        baseDir.resolve(".gradle/buildOutputCleanup/cache.properties").toString()
```
Comment (dorian-burihabwa-sonarsource) [inline integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java:358]: What is the motivation for analyzing this `buildOutputCleanup` directory? Is it a static, committed file or is an artifact introduced by gradle or on the task that ran before the sonar task?

**Review by dorian-burihabwa-sonarsource — CHANGES_REQUESTED** [review]
There are a few things we can tweak for clarity and consistency. I am not super comfortable with the large inclusion of hidden files and I am worried about the behavior of the scanner with folders such as scm folders or virtual environments

File: src/main/java/org/sonarqube/gradle/SourceCollector.java [hunk 4]
```diff
@@ -101,19 +112,27 @@ public SourceCollector(Set<Path> existingSources, Set<Path> directoriesToIgnore,
...
     return FileVisitResult.CONTINUE;
   }
 
-  private static boolean isHidden(Path path) {
-    return StreamSupport.stream(path.spliterator(), true)
-      .anyMatch(token -> token.toString().startsWith("."));
+  private boolean isHidden(Path path) {
+    Path relativePath = root.relativize(path);
+    while (relativePath != null && !relativePath.equals(root)) {
+      if (relativePath.toFile().isHidden()) {
```
Comment (alban-auzeill) [inline src/main/java/org/sonarqube/gradle/SourceCollector.java:126]: No need to convert the path to call `relativePath.toFile().isHidden()` you can do it using `Files.isHidden(path)`
But why changing `.startsWith(".")` by `.isHidden(path)`?
I don't know if it is reliable. If I extract a GitHub repo on a Linux and a Windows machine, will both OSs consider the same files hidden?

**Review by alban-auzeill — CHANGES_REQUESTED** [review]

File: src/main/java/org/sonarqube/gradle/SourceCollector.java [hunk 5]
```diff
@@ -81,6 +80,13 @@ public class SourceCollector implements FileVisitor<Path> {
     ".kt")).map(ext -> ext.toLowerCase(Locale.ROOT))
     .collect(Collectors.toSet());
 
+  private static final Set<String> INCLUDE_EXTENSIONS_FOR_HIDDEN_FILES = Set.of(".env", ".json", ".yml", ".yaml", ".properties", ".db", ".srcMainJava", ".py", ".xml", ".sh",
+    ".bash", ".ksh", ".zsh", ".bat", ".ps1", ".txt", ".config", ".settings", ".cnf");
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply src/main/java/org/sonarqube/gradle/SourceCollector.java:84]: Let's check if `srcMainJava` is part of the inclusions we want to add

File: integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java (hunk 3 above)
Comment (ADarko22) [reply integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java:358]: As discussed we'll exclude visiting the `.gradle` and `.git` directories

File: integrationTests/src/test/resources/multi-module-with-submodules/build.gradle.kts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 plugins {
   java
-  id("org.sonarqube") version "${version}"
+  id("org.sonarqube") version "5.1-SNAPSHOT"
```
Comment (alban-auzeill) [inline integrationTests/src/test/resources/multi-module-with-submodules/build.gradle.kts:3]: Oops. we need to revert this change!

**Review by dorian-burihabwa-sonarsource — APPROVED** [review]
